File: C#/DataStructures/DataStructures/stack/Queue.cs
using System;
using System.Collections;
using System.Collections.Generic;

namespace DataStructures.stack
{
    public class Queue<T> : IEnumerable<T>
    {
        private List<T> values = new List<T>();

        public int Size { get { return values.Count; } }
        public bool Empty { get { return values.Count == 0; } }

        public void Enqueue(T val)
        {
            values.Add(val);
        }

        public T Dequeue()
        {
            if(Empty)
            {
                Console.WriteLine("Warning: Attempt to dequeue from empty queue");
                return default(T);
            }

            T retval = values[0];
            values.RemoveAt(0);

            return retval;
        }

        public IEnumerator<T> GetEnumerator()
        {
            while (!Empty)
                yield return Dequeue();
        }

        IEnumerator IEnumerable.GetEnumerator()
        {
            while (!Empty)
                yield return Dequeue();
        }
    }
}

File: go/src/core/queue.go
package core

import "errors"

type Queue struct {
  data []int
}

func (q *Queue) Enqueue(item int) {
  q.data = append(q.data, item)
}

func (q *Queue) Dequeue() (int, error) {
  if len(q.data) == 0 { return -1, errors.New("attempted to pop from empty stack") }

  retval := q.data[0]
  q.data = q.data[1:]

  return retval, nil
}

func (q *Queue) Size() int {
  return len(q.data)
}
